cosmwasm/enclaves/shared/contract-engine/src/wasm/contract.rs

```diff
@@ -603,4 +654,677 @@ impl WasmiApi for ContractInstance {
+                    .collect::<Vec<&[u8]>>(),
+                pubkeys_data
+                    .iter()
+                    .map(|p| p.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+            )
+        } else if pubkeys_data.len() == 1 && messages_data.len() == signatures_data.len() {
+            // Replicate pubkey
+            (
+                messages_data
+                    .iter()
+                    .map(|m| m.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                signatures_data
+                    .iter()
+                    .map(|s| s.as_slice())
+                    .collect::<Vec<&[u8]>>(),
+                vec![pubkeys_data[0].as_slice()].repeat(signatures_data.len()),
+            )
+        } else {
+            debug!(
+                "ed25519_batch_verify() mismatched number of messages ({}) / signatures ({}) / public keys ({})", messages_data.len(),signatures_data.len(),pubkeys_data.len()
+            );
+
+            // return 5 == BatchErr
+            // https://github.com/CosmWasm/cosmwasm/blob/v1.0.0-beta5/packages/crypto/src/errors.rs#L97
+            return Ok(Some(RuntimeValue::I32(7)));
+        };
+
+        self.use_gas_externally(
+            (self.gas_costs.external_ed25519_verify
+                - self.gas_costs.external_ed25519_batch_verify_per_one) as u64 /* base cost in case signatures.len() == 0 */
+                + (signatures.len() as u64) * self.gas_costs.external_ed25519_batch_verify_per_one as u64,
+        )?;
+
+        let mut batch = ed25519_zebra::batch::Verifier::new();
+        for i in 0..signatures.len() {
+            let signature: ed25519_zebra::Signature = match ed25519_zebra::Signature::try_from(
+                signatures[i],
+            ) {
+                Ok(x) => x,
+                Err(err) => {
+                    debug!(
+                    "ed25519_batch_verify() failed to create an ed25519 signature from signatures[{}]: {:?}",
+                    i, err
+                );
+
+                    // return 4 == InvalidSignatureFormat
+                    // https://github.com/CosmWasm/cosmwasm/blob/v1.0.0-beta5/packages/crypto/src/errors.rs#L94
+                    return Ok(Some(RuntimeValue::I32(4)));
+                }
+            };
+
+            let pubkey: ed25519_zebra::VerificationKeyBytes =
+                match ed25519_zebra::VerificationKeyBytes::try_from(pubkeys[i]) {
+                    Ok(x) => x,
+                    Err(err) => {
+                        debug!(
+                        "ed25519_batch_verify() failed to create an ed25519 VerificationKey from public_keys[{}]: {:?}",
+                        i, err
+                    );
+
+                        // return 5 == InvalidPubkeyFormat
+                        // https://github.com/CosmWasm/cosmwasm/blob/v1.0.0-beta5/packages/crypto/src/errors.rs#L95
+                        return Ok(Some(RuntimeValue::I32(5)));
```

Comment:
This should come from the crypto package, and that way you can define an enum and use that instead of doing it like this